prusti-interface/src/specs/typed.rs

```diff
@@ -134,7 +136,71 @@ impl SpecificationItem<bool> {
     }
 }
 
+#[derive(Debug)]
+pub enum ProcedureSpecificationKindError {
+    /// Occurs whenever the relation between [ProcedureSpecificationKind]s is violated.
+    /// The relation is: predicates ⊂ pure functions ⊂ impure functions
+    /// This validation is not automatically performed during refinement. It needs to be invoked
+    /// manually with [SpecificationItem<ProcedureSpecificationKind>::validate]
+    InvalidSpecKindRefinement(ProcedureSpecificationKind, ProcedureSpecificationKind)
+}
+
+impl SpecificationItem<ProcedureSpecificationKind> {
+    pub fn is_pure(&self) -> Result<bool, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(matches!(self.extract_with_selective_replacement(),
+                Some(ProcedureSpecificationKind::Pure) | Some(ProcedureSpecificationKind::Predicate(_))))
+    }
+
+    pub fn is_impure(&self) -> Result<bool, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(match self.extract_with_selective_replacement() {
+            Some(refined) => matches!(refined, ProcedureSpecificationKind::Impure),
+            _ => true,
+        })
+    }
+
+    pub fn get_predicate_body(&self) -> Result<Option<&LocalDefId>, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(match self.extract_with_selective_replacement() {
+            Some(ProcedureSpecificationKind::Predicate(pred_id)) => Some(pred_id),
+            _ => None
+        })
+    }
+
+    /// Ensures that refined [ProcedureSpecificationKind]s are valid.
+    /// See [ProcedureSpecificationKindError] for detailed error descriptions.
+    pub fn validate(&self) -> Result<(), ProcedureSpecificationKindError> {
+        use ProcedureSpecificationKind::*;
+        if let SpecificationItem::Refined(base, refined) = self {
+            match (base, refined) {
+                // An impure method can be refined to anything
+                (Impure, _) |
+                // An pure method can be refined to pure method or a predicate
+                (Pure, Pure) |
+                (Pure, Predicate(_)) |
```

Comment:
Thinking about this some more: I don't think pure -> predicate or impure -> predicate should be allowed.

```rust
trait Foo        {             fn blah(&self) -> bool;                 }
impl Bar for Foo { predicate!{ fn blah(&self) -> bool { forall ... } } }
```

Here `Foo::blah` is impure and callable from non-spec code but a predicate should only be callable from specs. So a method like this might result in a call to the predicate from non-spec code:

```rust
fn client<T: Foo>(x: &T) { x.blah(); }
```

Reply:
Yeah that's right, this validation also needs to consider that no calls can happen to predicates. 
For your `client`, we would need to validate all callsites, which is not really practical and modular.
So we should only allow the following refinements:
- Impure -> {Impure, Pure}
- Pure -> Pure
- Predicate -> Predicate